From 8835cbb5e14fc7bed65eccd927f4bca2c712a45a Mon Sep 17 00:00:00 2001
From: Christian Hemp <c.hemp@phytec.de>
Date: Fri, 4 Oct 2013 09:03:09 +0200
Subject: [PATCH 24/55] imx6q: Add module revison detection

The pin SD4_DAT4 until SD4_DAT7 be used as revison control.
The pins will be internel pulled up so we read a 1111 for revison 1.
For revison two the first pin (bit) is pulled down (see schematic pfla-02
page 4 "SDIO, NAND-Flash".
On Module rev 1 the pins are connectet to the NAND but we have only 8bit NAND
also the i.MX6 only can handle 8bit NAND flashs.

Revisions:
	Rev 1:  0xF
	Rev 2:  0xE
		.
		.
		.
	Rev 15: 0x1
	Rev 16: 0x0

Signed-off-by: Christian Hemp <c.hemp@phytec.de>
---
 arch/arm/boards/phytec-pfla02/board.c  |   21 ++++++++++++++++++++-
 arch/arm/dts/imx6dl-phytec-pfla02.dtsi |    4 ++++
 arch/arm/dts/imx6q-phytec-pfla02.dtsi  |    4 ++++
 3 files changed, 28 insertions(+), 1 deletions(-)

diff --git a/arch/arm/boards/phytec-pfla02/board.c b/arch/arm/boards/phytec-pfla02/board.c
index c56d7a4..ed65836 100644
--- a/arch/arm/boards/phytec-pfla02/board.c
+++ b/arch/arm/boards/phytec-pfla02/board.c
@@ -34,6 +34,9 @@
 #include <mach/imx6.h>
 #include <mach/bbu.h>
 
+#define PHYFLEX_MODULE_REV_1	0x1
+#define PHYFLEX_MODULE_REV_2	0x2
+
 #define GPIO_2_11_PD_CTL	MX6_PAD_CTL_PUS_100K_DOWN | MX6_PAD_CTL_PUE | MX6_PAD_CTL_PKE | \
 				MX6_PAD_CTL_SPEED_MED | MX6_PAD_CTL_DSE_40ohm | MX6_PAD_CTL_HYS
 #define MX6Q_PAD_SD4_DAT3__GPIO_2_11_PD (_MX6Q_PAD_SD4_DAT3__GPIO_2_11 | MUX_PAD_CTRL(GPIO_2_11_PD_CTL))
@@ -50,6 +53,19 @@ static void phyflex_err006282_workaround(void)
 	gpio_direction_input(MX6_PHYFLEX_ERR006282);
 }
 
+
+static unsigned int get_module_rev(void)
+{
+	unsigned int val = 0;
+
+	val = gpio_get_value(IMX_GPIO_NR(2, 12));
+	val |= (gpio_get_value(IMX_GPIO_NR(2, 13)) << 1);
+	val |= (gpio_get_value(IMX_GPIO_NR(2, 14)) << 2);
+	val |= (gpio_get_value(IMX_GPIO_NR(2, 15)) << 3);
+
+	return 16 - val;
+}
+
 #define ETH_PHY_RST	IMX_GPIO_NR(3, 23)
 
 static int eth_phy_reset(void)
@@ -72,7 +88,8 @@ static void mmd_write_reg(struct phy_device *dev, int device, int reg, int val)
 
 static int ksz9031rn_phy_fixup(struct phy_device *dev)
 {
-	mmd_write_reg(dev, 2, 8, 0x039F);
+	if (get_module_rev() == PHYFLEX_MODULE_REV_2)
+		mmd_write_reg(dev, 2, 8, 0x039F);
 
 	return 0;
 }
@@ -94,6 +111,8 @@ static int phytec_pfla02_init(void)
 
 	phyflex_err006282_workaround();
 
+	printf("Module Revision: %u\n",get_module_rev());
+
 	if (of_machine_is_compatible("phytec,imx6q-pfla02")) {
 		flash_header_start = (void *)flash_header_phytec_pfla02_1gib_start;
 		flash_header_end = (void *)flash_header_phytec_pfla02_1gib_end;
diff --git a/arch/arm/dts/imx6dl-phytec-pfla02.dtsi b/arch/arm/dts/imx6dl-phytec-pfla02.dtsi
index 9279fb0..466e26d 100644
--- a/arch/arm/dts/imx6dl-phytec-pfla02.dtsi
+++ b/arch/arm/dts/imx6dl-phytec-pfla02.dtsi
@@ -50,6 +50,10 @@
 				MX6DL_PAD_EIM_D23__GPIO3_IO23    0x80000000
 				MX6DL_PAD_DISP0_DAT3__GPIO4_IO24 0x80000000 /* SPI NOR chipselect */
 				MX6DL_PAD_SD4_DAT3__GPIO2_IO11	 0x80000000
+				MX6DL_PAD_SD4_DAT4__GPIO2_IO12	0x80000000
+				MX6DL_PAD_SD4_DAT5__GPIO2_IO13	0x80000000
+				MX6DL_PAD_SD4_DAT6__GPIO2_IO14	0x80000000
+				MX6DL_PAD_SD4_DAT7__GPIO2_IO15	0x80000000
 			>;
 		};
 	};
diff --git a/arch/arm/dts/imx6q-phytec-pfla02.dtsi b/arch/arm/dts/imx6q-phytec-pfla02.dtsi
index 9fc767d..3694ea5 100644
--- a/arch/arm/dts/imx6q-phytec-pfla02.dtsi
+++ b/arch/arm/dts/imx6q-phytec-pfla02.dtsi
@@ -51,6 +51,10 @@
 				MX6Q_PAD_EIM_D23__GPIO3_IO23    0x80000000
 				MX6Q_PAD_DISP0_DAT3__GPIO4_IO24	0x80000000 /* SPI NOR chipselect */
 				MX6Q_PAD_SD4_DAT3__GPIO2_IO11	0x80000000
+				MX6Q_PAD_SD4_DAT4__GPIO2_IO12	0x80000000
+				MX6Q_PAD_SD4_DAT5__GPIO2_IO13	0x80000000
+				MX6Q_PAD_SD4_DAT6__GPIO2_IO14	0x80000000
+				MX6Q_PAD_SD4_DAT7__GPIO2_IO15	0x80000000
 			>;
 		};
 	};
-- 
1.7.0.4

